scipy/special/basic.py

```diff
@@ -2259,15 +2288,41 @@ def factorial(n, exact=False):
     >>> arr = np.array([3, 4, 5])
     >>> factorial(arr, exact=False)
     array([   6.,   24.,  120.])
+    >>> factorial(arr, exact=True)
+    array([  6,  24, 120])
     >>> factorial(5, exact=True)
     120L
 
     """
     if exact:
-        return math.factorial(n)
+        if np.ndim(n) == 0:
+            return math.factorial(n)
+        else:
+            n = asarray(n)
+            un = np.unique(n)
+
+            if any(abs(un.astype(int)) != un):
+                raise ValueError("Input must be non-negative integers for "
+                                 "'exact' mode.")
```

Comment:
This is, strictly speaking, a backcompat break. 

Also, while it's consistent with `math.factorial` above, it's inconsistent with the `exact=False` code path:

```
In [37]: factorial(-3)
Out[37]: array(0.0)

In [38]: factorial(-3, exact=True)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-38-178c9dd1705f> in <module>()
----> 1 factorial(-3, exact=True)

/home/br/repos/scipy/build/testenv/lib/python3.5/site-packages/scipy/special/basic.py in factorial(n, exact)
   2297     if exact:
   2298         if np.ndim(n) == 0:
-> 2299             return math.factorial(n)
   2300         else:
   2301             n = asarray(n)

ValueError: factorial() not defined for negative values
```

In the released scipy versions it was

```
In [6]: scipy.__version__
Out[6]: '0.17.0'

In [7]: factorial(-3)
Out[7]: array(0.0)

In [8]: factorial(-3, exact=True)
Out[8]: 0
```

Also consider this

```
In [41]: factorial([2, -3], exact=True)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-41-d1260e99c4d0> in <module>()
----> 1 factorial([2, -3], exact=True)

/home/br/repos/scipy/build/testenv/lib/python3.5/site-packages/scipy/special/basic.py in factorial(n, exact)
   2303 
   2304             if any(abs(un.astype(int)) != un):
-> 2305                 raise ValueError("Input must be non-negative integers for "
   2306                                  "'exact' mode.")
   2307 

ValueError: Input must be non-negative integers for 'exact' mode.
```

I think this is inconsistent with how pretty much the whole rest of scipy.special deals with arrays where some elements are incorrect.

All in all I mildly prefer to declare the ship as having sailed and restore the backwards compatibility.